### PR TITLE
OSD: during test start, not all osds started due to consum map hang

### PR DESCRIPTION
**Pull request by NitzanMordhai**
The OSD log won't show many messages during the consume map, moving OSDmap services from header files and adding more debug messages.

Fixes: https://tracker.ceph.com/issues/57699
Signed-off-by: Nitzan Mordechai <nmordec@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by rzarzynski — COMMENTED** [review]
1. I'm not sure we can say `Fixes` here; the change is a debug facility. On the other hand it's clear we would love to backport it to increase the chance for replication. I would at least mention explicitly in the commit message, perhaps even with CAPS LOCK, this isn't the fix.
2. How about dissecting the code shuffling into a dedicated commit?

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -352,12 +352,18 @@ void OSDService::identify_splits_and_merges(
   set<pair<spg_t,epoch_t>> *split_children,
```
Comment (rzarzynski) [inline src/osd/OSD.cc:352]: >  OSD: during test start, not all osds started due to consum map hang 

I think there is a typo in first's commit title: `consum map` -> `consume_map`.

File: src/osd/OSD.h [hunk 2]
```diff
@@ -197,40 +197,11 @@ class OSDService : public Scrub::ScrubSchedListener {
...
-    std::lock_guard l(pre_publish_lock);
-    epoch_t e = next_osdmap->get_epoch();
-    std::map<epoch_t, unsigned>::iterator i =
-      map_reservations.insert(std::make_pair(e, 0)).first;
-    i->second++;
-    return next_osdmap;
-  }
+  OSDMapRef get_nextmap_reserved();
   /// releases reservation on map
-  void release_map(OSDMapRef osdmap) {
```
Comment (rzarzynski) [inline src/osd/OSD.h:209]: Does the first commit contain any fix or it's just about shuffling the code + extra debugs?   If the later, saying this explicitly in the commit description would be helpful.

Comment (NitzanMordhai) [reply src/osd/OSD.h:209]: Done

File: src/osd/OSD.cc (hunk 1 above)
Comment (NitzanMordhai) [reply src/osd/OSD.cc:352]: Done

**Review by sseshasa — APPROVED** [review]
As discussed, acquiring the shard lock before updating the configuration in mClock is not necessary. Based on this, I am approving the change.

From the logs, I don't clearly see the deadlock due to the shard lock which is being removed. But running tests with additional logging, the reason for the hang should become clearer. I also updated the related tracker with details about other trackers where we have been seeing similar symptoms in the past. May be all these trackers can be clubbed together once we RCA this.

**Comment by NitzanMordhai** [comment]
jenkins test api

**Comment by NitzanMordhai** [comment]
jenkins test api

**Comment by ljflores** [comment]
Rados suite review: https://pulpito.ceph.com/?branch=wip-yuri-testing-2023-01-23-0926

Failures, unrelated:
    1. https://tracker.ceph.com/issues/58587 -- new tracker
    2. https://tracker.ceph.com/issues/58585
    3. https://tracker.ceph.com/issues/58098 -- fix merged to latest main
    4. https://tracker.ceph.com/issues/58256 -- fix merged to latest main
    5. https://tracker.ceph.com/issues/57900
    6. https://tracker.ceph.com/issues/58475
    7. https://tracker.ceph.com/issues/58560

Details:
    1. test_dedup_tool.sh: test_dedup_object fails when pool 'dedup_chunk_pool' does not exist - Ceph - RADOS
    2. rook: failed to pull kubelet image - Ceph - Orchestrator
    3. qa/workunits/rados/test_crash.sh: crashes are never posted - Ceph - RADOS
    4. ObjectStore/StoreTestSpecificAUSize.SpilloverTest/2: Expected: (logger->get(l_bluefs_slow_used_bytes)) >= (16 * 1024 * 1024), actual: 0 vs 16777216 - Ceph - RADOS
    5. mon/crush_ops.sh: mons out of quorum - Ceph - RADOS
    6. test_dashboard_e2e.sh: Conflicting peer dependency: postcss@8.4.21 - Ceph - Mgr - Dashboard
    7. test_envlibrados_for_rocksdb.sh failed to subscrib repo - Ceph

**Pull request merged**